src/prefect/task_engine.py

```diff
@@ -716,6 +725,22 @@ def start(
     def transaction_context(self) -> Generator[Transaction, None, None]:
         result_factory = getattr(TaskRunContext.get(), "result_factory", None)
 
+        def __prefect_set_task_state_to_rolled_back(txn: Transaction):
+            if txn.state != TransactionState.STAGED:
```

Comment:
See prior discussion [here](https://github.com/PrefectHQ/prefect/pull/14662#discussion_r1688712832).

This short-circuit is primarily so that a task like:
```python
@task
def i_fail():
    raise ValueError('whoops')
```
doesn't now transition through a Completed(name="RolledBack") state before retrying and/or failing. The task should only be in this state if it had previously completed, but was then rolled back and its rollback hooks run.